### PR TITLE
Certification

### PR DESCRIPTION
**Pull request by danieldanan**
Replacing Performance Center user and password fields + Proxy user and password fields by two Credentials fields.

### CODE REVIEW

File: src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java [hunk 1]
```diff
@@ -484,96 +681,119 @@ private boolean isPluginActive(String pluginDisplayName){
...
-        saveFileToWorkspacePath(pcClient,pcModel.getTrendReportId(true),runId,TrendReportTypes.DataType.Transaction, TrendReportTypes.PctType.TRT, TrendReportTypes.Measurement.PCT_MEDIAN);
-        saveFileToWorkspacePath(pcClient,pcModel.getTrendReportId(true),runId,TrendReportTypes.DataType.Transaction, TrendReportTypes.PctType.TRT, TrendReportTypes.Measurement.PCT_STDDEVIATION);
-        saveFileToWorkspacePath(pcClient,pcModel.getTrendReportId(true),runId,TrendReportTypes.DataType.Transaction, TrendReportTypes.PctType.TRT, TrendReportTypes.Measurement.PCT_COUNT1);
-        //saveFileToWorkspacePath(pcClient,pcModel.getTrendReportId(true),runId,TrendReportTypes.DataType.Transaction, TrendReportTypes.PctType.TRT, TrendReportTypes.Measurement.PCT_SUM1);
-        saveFileToWorkspacePath(pcClient,pcModel.getTrendReportId(true),runId,TrendReportTypes.DataType.Transaction, TrendReportTypes.PctType.TRT, TrendReportTypes.Measurement.PCT_PERCENTILE_90);
+        saveFileToWorkspacePath(pcClient,getPcModel().getTrendReportId(true),runId,TrendReportTypes.DataType.Transaction, TrendReportTypes.PctType.TRT, TrendReportTypes.Measurement.PCT_MINIMUM);
```
[the remark this answers is not in the document]
Comment (danieldanan) [reply src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java]: thanks. array created

File: src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java [hunk 2]
```diff
@@ -484,96 +681,119 @@ private boolean isPluginActive(String pluginDisplayName){
 
     private void updateCSVFilesForPlot(PcClient pcClient, int runId) throws IOException, PcException, IntrospectionException, NoSuchMethodException {
 
-        //Map<String, String> measurementMap =pcClient.getTrendReportByXML(pcModel.getTrendReportId(true),runId,TrendReportTypes.DataType.Transaction, TrendReportTypes.PctType.TRT, TrendReportTypes.Measurement.PCT_AVERAGE);
+        //Map<String, String> measurementMap =pcClient.getTrendReportByXML(getPcModel().getTrendReportId(true),runId,TrendReportTypes.DataType.Transaction, TrendReportTypes.PctType.TRT, TrendReportTypes.Measurement.PCT_AVERAGE);
```
[the remark this answers is not in the document]
Comment (danieldanan) [reply src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java]: done

File: src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java [hunk 3]
```diff
@@ -484,96 +681,119 @@ private boolean isPluginActive(String pluginDisplayName){
...
+                PrintWriter writer = new PrintWriter(file);
+                for (String key : measurementMap.keySet()) {
+                    writer.print(key + ",");
+                }
+                writer.print("\r\n");
+                for (String value : measurementMap.values()) {
+                    writer.print(value + ",");
+                }
+                writer.close();
+                //     logger.println(String.format("%s - %s Created.", simpleDateFormater(), fileName);
```
[the remark this answers is not in the document]
Comment (danieldanan) [reply src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java]: done

File: src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java [hunk 4]
```diff
@@ -803,8 +1032,21 @@ public boolean isStatusBySLA() {
     }
 
     public String getProxyOutURL(){ return getPcModel().getProxyOutURL();}
-    public String getProxyOutUser(){ return getPcModel().getProxyOutUser();}
-    public String getProxyOutPassword(){ return getPcModel().getProxyOutPassword();}
+
+    private String  simpleDateFormater()
```
[the remark this answers is not in the document]
Comment (danieldanan) [reply src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java]: created a class.

File: src/main/java/com/microfocus/application/automation/tools/pc/PcClient.java [hunk 5]
```diff
@@ -50,24 +51,33 @@
...
 
     public PcClient(PcModel pcModel, PrintStream logger) {
         try {
             model = pcModel;
-
-            if(model.getProxyOutURL(true) != null && !model.getProxyOutURL(true).isEmpty()){
-                logger.println(String.format("%s - Using proxy: %s", _simpleDateFormat.format(new Date()), model.getProxyOutURL(true)));
+            String credentialsProxyId = model.getCredentialsProxyId(true);
+            usernamePCPasswordCredentialsForProxy = PcBuilder.getCredentialsId(credentialsProxyId);
+            String proxyOutUser = (usernamePCPasswordCredentialsForProxy == null || model.getProxyOutURL(true).isEmpty()) ? "" : usernamePCPasswordCredentialsForProxy.getUsername();
```
[the remark this answers is not in the document]
Comment (danieldanan) [reply src/main/java/com/microfocus/application/automation/tools/pc/PcClient.java]: no duplicated code found

File: src/main/java/com/microfocus/application/automation/tools/pc/PcClient.java [hunk 6]
```diff
@@ -80,17 +90,28 @@ public PcClient(PcModel pcModel, PrintStream logger) {
...
+                loggedIn = restProxy.authenticate(usernamePCPasswordCredentials.getUsername(), usernamePCPasswordCredentials.getPassword().getPlainText());
+            }
+            else {
+                logger.println(String.format("%s - Trying to login\n[PCServer='%s://%s', User='%s']", simpleDateFormater(), model.isHTTPSProtocol(), model.getPcServerName(true), PcBuilder.usernamePCPasswordCredentials.getUsername()));
+                loggedIn = restProxy.authenticate(PcBuilder.usernamePCPasswordCredentials.getUsername(), PcBuilder.usernamePCPasswordCredentials.getPassword().getPlainText());
+            }
         } catch (PcException e) {
-            logger.println(String.format("%s - %s", _simpleDateFormat.format(new Date()), e.getMessage()));
+            logger.println(String.format("%s - %s", simpleDateFormater(), e.getMessage()));
           //  stackTraceToString(e);
```
[the remark this answers is not in the document]
Comment (danieldanan) [reply src/main/java/com/microfocus/application/automation/tools/pc/PcClient.java]: done

File: src/main/java/com/microfocus/application/automation/tools/pc/PcClient.java [hunk 7]
```diff
@@ -108,50 +129,157 @@ public int startRun() throws NumberFormatException, ClientProtocolException, PcE
...
+            response = restProxy.startRun(testID,
+                    testInstance,
+                    model.getTimeslotDuration(),
+                    model.getPostRunAction().getValue(),
+                    model.isVudsMode());
+            logger.println(String.format("%s - Run started (TestID: %s, RunID: %s, TimeslotID: %s)", simpleDateFormater(),
+                    response.getTestID(), response.getID(), response.getTimeslotID()));
+            return response.getID();
+        }
+        catch (NumberFormatException ex) {
```
[the remark this answers is not in the document]
Comment (danieldanan) [reply src/main/java/com/microfocus/application/automation/tools/pc/PcClient.java]: grouped NumberFormatException|ClientProtocolException|PcException  in one catch

File: src/main/java/com/microfocus/application/automation/tools/pc/PcClient.java [hunk 8]
```diff
@@ -108,50 +129,157 @@ public int startRun() throws NumberFormatException, ClientProtocolException, PcE
...
+            logger.println(String.format("%s - StartRun failed. Error: %s", simpleDateFormater(), ex.getMessage()));
+        }
+        if (!("RETRY".equals(model.getRetry()))) {
+            return 0;
+        }
+        else {
+            //counter
+            int retryCount = 0;
+            //values
+            int retryDelay = verifyStringValueIsIntAndPositive("retryDelay", model.getRetryDelay(), 5);
```
[the remark this answers is not in the document]
Comment (danieldanan) [reply src/main/java/com/microfocus/application/automation/tools/pc/PcClient.java]: indeed. function moved and moved to PCModel (and variables' type change in PCModel)

File: src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java [hunk 9]
```diff
@@ -656,12 +622,12 @@ private Testsuites parsePcTrendResponse(Testsuites ret,Run<?, ?> build,PcClient
...
             // this helps to show the transaction of each result
             if (isPluginActive("Plot plugin")) {
-                logger.println(String.format("%s Updating csv files for Trending Charts.", simpleDateFormater()));
+                logger.println(String.format("%s Updating csv files for Trending Charts.", dateFormatter.getDate()));
                 updateCSVFilesForPlot(pcClient, runID);
                 String plotUrlPath = "/job/" + build.getParent().getName() + "/plot";
-                logger.println(String.format("%s - %s",simpleDateFormater(), HyperlinkNote.encodeTo(plotUrlPath, "Trending Charts"))); // + HyperlinkNote.encodeTo("https://wiki.jenkins-ci.org/display/JENKINS/HP+Application+Automation+Tools#HPApplicationAutomationTools-RunningPerformanceTestsusingHPPerformanceCenter","More Info"));
+                logger.println(String.format("%s - %s",dateFormatter.getDate(), HyperlinkNote.encodeTo(plotUrlPath, "Trending Charts"))); // + HyperlinkNote.encodeTo("https://wiki.jenkins-ci.org/display/JENKINS/HP+Application+Automation+Tools#HPApplicationAutomationTools-RunningPerformanceTestsusingHPPerformanceCenter","More Info"));
```
Comment (grount) [inline src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java]: The wiki is no longer correct on this one as well

File: src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java [hunk 10]
```diff
@@ -83,30 +95,49 @@
     public static final String    PENDING         = "Pending";
     public static final String    PUBLISHING      = "Publishing";
     public static final String    ERROR           = "Error";
-    
-    private PcModel pcModel;
 
+    private PcModel pcModel;
+    public static UsernamePasswordCredentials usernamePCPasswordCredentials;
+    public static UsernamePasswordCredentials usernamePCPasswordCredentialsForProxy;
+    private static Run<?, ?> _run;
```
Comment (grount) [inline src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java]: Can this be transient?

File: src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java [hunk 11]
```diff
@@ -83,30 +95,49 @@
...
     private String testName;
     private FilePath pcReportFile;
     private String junitResultsFileName;
-    private PrintStream logger;
+    private static PrintStream logger;
     private File WorkspacePath;
-    private SimpleDateFormat _simpleDateFormat = new SimpleDateFormat ("E MM.dd.yyyy 'at' hh:mm:ss a zzz");
-
-
+    private AbstractBuild<?, ?> _build;
```
Comment (grount) [inline src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java]: Can this be transient also?

File: src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java (hunk 10 above)
Comment (danieldanan) [reply src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java]: done

File: src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java (hunk 11 above)
Comment (danieldanan) [reply src/main/java/com/microfocus/application/automation/tools/run/PcBuilder.java]: done

**Comment by grount** [comment]
Approved for test build

**Pull request merged**